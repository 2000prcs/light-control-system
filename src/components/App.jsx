import React, { Component } from 'react';
import Nav from './Nav.jsx';
import Control from './Control.jsx';
import Display from './Display.jsx';
import Menu from './Menu.jsx';

// Sample data for testing without Light API
import { sampleData } from '../../public/lib/sampleData.js';

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: sampleData,
      currentRoom: {},
      userData: {
        userName: 'Moyeong Lee',
      },
    };

    this.getRoomInfo = this.getRoomInfo.bind(this);
    this.getCurrentRoom = this.getCurrentRoom.bind(this);
    this.lightControl = this.lightControl.bind(this);
    this.getSwitchStatus = this.getSwitchStatus.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  // Send GET request to server when the component is mounted
  componentDidMount() {
    this.getRoomInfo();
  }

  // Fetch room data from Light API
  getRoomInfo() {
    fetch('http://localhost:3000/api/v1/device')
      .then(response => response.json())
      .then((result) => {
        this.setState({ data: result.data });
      })
      .catch(errors => console.error(errors));
  }

  // Hightlight the room and display its controller when user selects the room
  getCurrentRoom(info) {
    this.setState({ currentRoom: info });
  }

  // Check if light switch is on or off
  getSwitchStatus(value) {
    const room = this.state.currentRoom;
    room.active = value;
    if (!value) {
      room.brightness = 0;
    }
    this.setState({ currentRoom: room });
  }

  // Change brightness value dynamically
  lightControl(value) {
    if (value === 0) {
      this.getSwitchStatus(false);
    }
    const room = this.state.currentRoom;
    room.brightness = Math.ceil(value);
    this.setState({ currentRoom: room });
  }

  // Change room name as user types
  handleChange(e) {
    const currentRoomNameNode = document.querySelector('.highlight a span b');
    if (currentRoomNameNode) {
      currentTableNameNode.innerText = e.target.value;
    }
    const room = this.state.currentRoom;
    room.name = e.target.value;
    this.setState({ currentRoom: room });
  }

  render() {
    const { data, currentRoom, userData } = this.state;

    return (
      <div className="container">
        <Nav userData={userData} />
        <Menu room={currentRoom} handleChange={this.handleChange} />
        <div className="main">
          <div className="control">
            <Control room={data} getCurrentRoom={this.getCurrentRoom} getSwitchStatus={this.getSwitchStatus} />
          </div>
          <div className="display">
            <Display room={currentRoom} lightControl={this.lightControl} />
          </div>
        </div>
      </div>
    );
  }
}
