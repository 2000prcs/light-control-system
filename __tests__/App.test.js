import React from 'react';
import sinon from 'sinon';
import "isomorphic-fetch";
import App from '../src/components/App.jsx';

let roomData = {
  active: false,
  brightness: 50,
  id: 1,
  name: 'Balcony',
};


describe('App component', () => {
  test('Should render App', () => {
    const AppWrapper = shallow(<App />);
    expect(AppWrapper).toMatchSnapshot();
  });

  test('Should fetch room information from API upon mounting', () => {
    sinon.spy(App.prototype, 'getRoomInfo');
    const AppWrapper = mount(<App />);
    expect(App.prototype.getRoomInfo.calledOnce).toBe(true);
    App.prototype.getRoomInfo.restore();
    AppWrapper.unmount();
  });

  test('Should get a current room info from Controller', () => {
    const AppWrapper = mount(<App />);
    AppWrapper.setState({ currentRoom: '' });
    AppWrapper.instance().getCurrentRoom({
      id: 4,
      name: 'Entrance',
      active: true,
      brightness: 100,
    });
    AppWrapper.update();
    const currentRoomId = AppWrapper.state().currentRoom.id;
    expect(currentRoomId).toBe(4);
    AppWrapper.unmount();
  });

  test('Should turn a switch on when getSwitchStatus is invoked with true value', () => {
    const AppWrapper = mount(<App />);
    AppWrapper.setState({ currentRoom: roomData });
    AppWrapper.instance().getSwitchStatus(true);
    AppWrapper.update();
    const currentRoomId = AppWrapper.state().currentRoom.active;
    expect(currentRoomId).toBe(true);
    AppWrapper.unmount();
  });

  test('Should set the room\'s brightness to 0 when getSwitchStatus is invoked with false value', () => {
    const AppWrapper = mount(<App />);
    AppWrapper.setState({ currentRoom: roomData });
    AppWrapper.instance().getSwitchStatus(false);
    AppWrapper.update();
    const currentRoomId = AppWrapper.state().currentRoom.brightness;
    expect(currentRoomId).toBe(0);
    AppWrapper.unmount();
  });

  test('Should set a brightness when lightControl is invoked', () => {
    const AppWrapper = mount(<App />);
    AppWrapper.setState({ currentRoom: roomData });
    AppWrapper.instance().lightControl(77.77);
    AppWrapper.update();
    const currentRoomId = AppWrapper.state().currentRoom.brightness;
    expect(currentRoomId).toBe(78);
    AppWrapper.unmount();
  });

  test('Should change the selected room name as a user types in the input', () => {
    const AppWrapper = mount(<App />);
    AppWrapper.setState({ currentRoom: roomData });
    AppWrapper.find('input').simulate('change', { target: { value: 'karaoke' } }); // Mock event object, for testing purposes
    AppWrapper.update();
    expect(AppWrapper.state().currentRoom.name).toBe('karaoke');
    AppWrapper.unmount();
  });
});
