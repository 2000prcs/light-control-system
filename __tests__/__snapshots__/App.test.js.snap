// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Should render App 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Nav
          userData={
            Object {
              "userName": "Moyeong Lee",
            }
          }
        />,
        <Display
          room={Object {}}
        />,
        <div
          className="main"
        >
          <div
            className="control"
          >
            <Control
              getCurrentRoom={[Function]}
              getSwitchStatus={[Function]}
              room={
                Array [
                  Object {
                    "active": true,
                    "brightness": 77,
                    "id": 1,
                    "name": "Mo's room",
                  },
                  Object {
                    "active": true,
                    "brightness": 100,
                    "id": 2,
                    "name": "Mo's living room",
                  },
                  Object {
                    "active": false,
                    "brightness": 0,
                    "id": 3,
                    "name": "Mo's restroom",
                  },
                  Object {
                    "active": true,
                    "brightness": 40,
                    "id": 4,
                    "name": "Mo's game room",
                  },
                  Object {
                    "active": false,
                    "brightness": 20,
                    "id": 5,
                    "name": "Mo's karaoke room",
                  },
                  Object {
                    "active": true,
                    "brightness": 80,
                    "id": 6,
                    "name": "Mo's piano room",
                  },
                  Object {
                    "active": true,
                    "brightness": 95,
                    "id": 7,
                    "name": "Mo's library",
                  },
                  Object {
                    "active": false,
                    "brightness": 66,
                    "id": 8,
                    "name": "Mo's secret room",
                  },
                ]
              }
            />
          </div>
          <div
            className="display"
          >
            <Display
              lightControl={[Function]}
              room={Object {}}
            />
          </div>
        </div>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "userData": Object {
            "userName": "Moyeong Lee",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "room": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="control"
            >
              <Control
                getCurrentRoom={[Function]}
                getSwitchStatus={[Function]}
                room={
                  Array [
                    Object {
                      "active": true,
                      "brightness": 77,
                      "id": 1,
                      "name": "Mo's room",
                    },
                    Object {
                      "active": true,
                      "brightness": 100,
                      "id": 2,
                      "name": "Mo's living room",
                    },
                    Object {
                      "active": false,
                      "brightness": 0,
                      "id": 3,
                      "name": "Mo's restroom",
                    },
                    Object {
                      "active": true,
                      "brightness": 40,
                      "id": 4,
                      "name": "Mo's game room",
                    },
                    Object {
                      "active": false,
                      "brightness": 20,
                      "id": 5,
                      "name": "Mo's karaoke room",
                    },
                    Object {
                      "active": true,
                      "brightness": 80,
                      "id": 6,
                      "name": "Mo's piano room",
                    },
                    Object {
                      "active": true,
                      "brightness": 95,
                      "id": 7,
                      "name": "Mo's library",
                    },
                    Object {
                      "active": false,
                      "brightness": 66,
                      "id": 8,
                      "name": "Mo's secret room",
                    },
                  ]
                }
              />
            </div>,
            <div
              className="display"
            >
              <Display
                lightControl={[Function]}
                room={Object {}}
              />
            </div>,
          ],
          "className": "main",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Control
                getCurrentRoom={[Function]}
                getSwitchStatus={[Function]}
                room={
                  Array [
                    Object {
                      "active": true,
                      "brightness": 77,
                      "id": 1,
                      "name": "Mo's room",
                    },
                    Object {
                      "active": true,
                      "brightness": 100,
                      "id": 2,
                      "name": "Mo's living room",
                    },
                    Object {
                      "active": false,
                      "brightness": 0,
                      "id": 3,
                      "name": "Mo's restroom",
                    },
                    Object {
                      "active": true,
                      "brightness": 40,
                      "id": 4,
                      "name": "Mo's game room",
                    },
                    Object {
                      "active": false,
                      "brightness": 20,
                      "id": 5,
                      "name": "Mo's karaoke room",
                    },
                    Object {
                      "active": true,
                      "brightness": 80,
                      "id": 6,
                      "name": "Mo's piano room",
                    },
                    Object {
                      "active": true,
                      "brightness": 95,
                      "id": 7,
                      "name": "Mo's library",
                    },
                    Object {
                      "active": false,
                      "brightness": 66,
                      "id": 8,
                      "name": "Mo's secret room",
                    },
                  ]
                }
              />,
              "className": "control",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "getCurrentRoom": [Function],
                "getSwitchStatus": [Function],
                "room": Array [
                  Object {
                    "active": true,
                    "brightness": 77,
                    "id": 1,
                    "name": "Mo's room",
                  },
                  Object {
                    "active": true,
                    "brightness": 100,
                    "id": 2,
                    "name": "Mo's living room",
                  },
                  Object {
                    "active": false,
                    "brightness": 0,
                    "id": 3,
                    "name": "Mo's restroom",
                  },
                  Object {
                    "active": true,
                    "brightness": 40,
                    "id": 4,
                    "name": "Mo's game room",
                  },
                  Object {
                    "active": false,
                    "brightness": 20,
                    "id": 5,
                    "name": "Mo's karaoke room",
                  },
                  Object {
                    "active": true,
                    "brightness": 80,
                    "id": 6,
                    "name": "Mo's piano room",
                  },
                  Object {
                    "active": true,
                    "brightness": 95,
                    "id": 7,
                    "name": "Mo's library",
                  },
                  Object {
                    "active": false,
                    "brightness": 66,
                    "id": 8,
                    "name": "Mo's secret room",
                  },
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Display
                lightControl={[Function]}
                room={Object {}}
              />,
              "className": "display",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "lightControl": [Function],
                "room": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Nav
            userData={
              Object {
                "userName": "Moyeong Lee",
              }
            }
          />,
          <Display
            room={Object {}}
          />,
          <div
            className="main"
          >
            <div
              className="control"
            >
              <Control
                getCurrentRoom={[Function]}
                getSwitchStatus={[Function]}
                room={
                  Array [
                    Object {
                      "active": true,
                      "brightness": 77,
                      "id": 1,
                      "name": "Mo's room",
                    },
                    Object {
                      "active": true,
                      "brightness": 100,
                      "id": 2,
                      "name": "Mo's living room",
                    },
                    Object {
                      "active": false,
                      "brightness": 0,
                      "id": 3,
                      "name": "Mo's restroom",
                    },
                    Object {
                      "active": true,
                      "brightness": 40,
                      "id": 4,
                      "name": "Mo's game room",
                    },
                    Object {
                      "active": false,
                      "brightness": 20,
                      "id": 5,
                      "name": "Mo's karaoke room",
                    },
                    Object {
                      "active": true,
                      "brightness": 80,
                      "id": 6,
                      "name": "Mo's piano room",
                    },
                    Object {
                      "active": true,
                      "brightness": 95,
                      "id": 7,
                      "name": "Mo's library",
                    },
                    Object {
                      "active": false,
                      "brightness": 66,
                      "id": 8,
                      "name": "Mo's secret room",
                    },
                  ]
                }
              />
            </div>
            <div
              className="display"
            >
              <Display
                lightControl={[Function]}
                room={Object {}}
              />
            </div>
          </div>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "userData": Object {
              "userName": "Moyeong Lee",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "room": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="control"
              >
                <Control
                  getCurrentRoom={[Function]}
                  getSwitchStatus={[Function]}
                  room={
                    Array [
                      Object {
                        "active": true,
                        "brightness": 77,
                        "id": 1,
                        "name": "Mo's room",
                      },
                      Object {
                        "active": true,
                        "brightness": 100,
                        "id": 2,
                        "name": "Mo's living room",
                      },
                      Object {
                        "active": false,
                        "brightness": 0,
                        "id": 3,
                        "name": "Mo's restroom",
                      },
                      Object {
                        "active": true,
                        "brightness": 40,
                        "id": 4,
                        "name": "Mo's game room",
                      },
                      Object {
                        "active": false,
                        "brightness": 20,
                        "id": 5,
                        "name": "Mo's karaoke room",
                      },
                      Object {
                        "active": true,
                        "brightness": 80,
                        "id": 6,
                        "name": "Mo's piano room",
                      },
                      Object {
                        "active": true,
                        "brightness": 95,
                        "id": 7,
                        "name": "Mo's library",
                      },
                      Object {
                        "active": false,
                        "brightness": 66,
                        "id": 8,
                        "name": "Mo's secret room",
                      },
                    ]
                  }
                />
              </div>,
              <div
                className="display"
              >
                <Display
                  lightControl={[Function]}
                  room={Object {}}
                />
              </div>,
            ],
            "className": "main",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Control
                  getCurrentRoom={[Function]}
                  getSwitchStatus={[Function]}
                  room={
                    Array [
                      Object {
                        "active": true,
                        "brightness": 77,
                        "id": 1,
                        "name": "Mo's room",
                      },
                      Object {
                        "active": true,
                        "brightness": 100,
                        "id": 2,
                        "name": "Mo's living room",
                      },
                      Object {
                        "active": false,
                        "brightness": 0,
                        "id": 3,
                        "name": "Mo's restroom",
                      },
                      Object {
                        "active": true,
                        "brightness": 40,
                        "id": 4,
                        "name": "Mo's game room",
                      },
                      Object {
                        "active": false,
                        "brightness": 20,
                        "id": 5,
                        "name": "Mo's karaoke room",
                      },
                      Object {
                        "active": true,
                        "brightness": 80,
                        "id": 6,
                        "name": "Mo's piano room",
                      },
                      Object {
                        "active": true,
                        "brightness": 95,
                        "id": 7,
                        "name": "Mo's library",
                      },
                      Object {
                        "active": false,
                        "brightness": 66,
                        "id": 8,
                        "name": "Mo's secret room",
                      },
                    ]
                  }
                />,
                "className": "control",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "getCurrentRoom": [Function],
                  "getSwitchStatus": [Function],
                  "room": Array [
                    Object {
                      "active": true,
                      "brightness": 77,
                      "id": 1,
                      "name": "Mo's room",
                    },
                    Object {
                      "active": true,
                      "brightness": 100,
                      "id": 2,
                      "name": "Mo's living room",
                    },
                    Object {
                      "active": false,
                      "brightness": 0,
                      "id": 3,
                      "name": "Mo's restroom",
                    },
                    Object {
                      "active": true,
                      "brightness": 40,
                      "id": 4,
                      "name": "Mo's game room",
                    },
                    Object {
                      "active": false,
                      "brightness": 20,
                      "id": 5,
                      "name": "Mo's karaoke room",
                    },
                    Object {
                      "active": true,
                      "brightness": 80,
                      "id": 6,
                      "name": "Mo's piano room",
                    },
                    Object {
                      "active": true,
                      "brightness": 95,
                      "id": 7,
                      "name": "Mo's library",
                    },
                    Object {
                      "active": false,
                      "brightness": 66,
                      "id": 8,
                      "name": "Mo's secret room",
                    },
                  ],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Display
                  lightControl={[Function]}
                  room={Object {}}
                />,
                "className": "display",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "lightControl": [Function],
                  "room": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
